astropy/coordinates/representation.py

```diff
@@ -1593,6 +1618,55 @@ def represent_as(self, other_class, differential_class=None):
 
         return super().represent_as(other_class, differential_class)
 
+    def transform(self, matrix):
+        r"""Transform the unit-spherical coordinates using a 3x3 matrix.
+
+        This returns a new representation and does not modify the original one.
+        Any differentials attached to this representation will also be
+        transformed.
+
+        Parameters
+        ----------
+        matrix : (3,3) array-like
+            A 3x3 transformation matrix, such as a rotation matrix.
+
+        Returns
+        -------
+        `UnitSphericalRepresentation` or `SphericalRepresentation`
+            If ``matrix`` is O(3) -- :math:`M \dot M^T = I` -- like a rotation,
+            then the result is a `UnitSphericalRepresentation`.
+            The differentials' classes are also preserved.
+            All other matrices will change the distance, so the dimensional
+            representation is used instead.
+            Likewise, the classes of the differentials will be made dimensional.
+
+        """
+        # the transformation matrix does not need to be a rotation matrix,
+        # so the unit-distance is not guaranteed. For speed, we check if the
+        # matrix is in O(3) (rotations, proper and improper) and keeps lengths.
+        if is_rotation(matrix, allow_improper=True):  # remain in unit-rep
+            if self.differentials:
+                # TODO! shortcut if there are differentials.
+                # Currently just super, which uses Cartesian backend.
+                rep = super().transform(matrix)
+
+            else:
+                xyz = erfa_ufunc.s2c(self.lon, self.lat)
+                p = erfa_ufunc.rxp(matrix, xyz)
+                lon, lat, _ = erfa_ufunc.p2s(p)
```

Comment:
I'd use `c2s` instead, to avoid erfa calculating the size.

Reply:
✅